fineract-provider/src/main/java/org/apache/fineract/portfolio/savings/service/SavingsApplicationProcessWritePlatformServiceJpaRepositoryImpl.java

```diff
@@ -86,470 +93,715 @@
+          this.savingsProductRepository = savingsProductRepository;
+          this.noteRepository = noteRepository;
+          this.staffRepository = staffRepository;
+          this.savingsAccountApplicationTransitionApiJsonValidator = savingsAccountApplicationTransitionApiJsonValidator;
+          this.savingsAccountChargeAssembler = savingsAccountChargeAssembler;
+          this.commandProcessingService = commandProcessingService;
+          this.savingsAccountDomainService = savingsAccountDomainService;
+          this.accountNumberFormatRepository = accountNumberFormatRepository;
+          this.savingsAccountWritePlatformService = savingsAccountWritePlatformService;
+          this.businessEventNotifierService = businessEventNotifierService;
+          this.entityDatatableChecksWritePlatformService = entityDatatableChecksWritePlatformService;
+          this.gsimRepository = gsimRepository;
+          this.groupRepositoryWrapper = groupRepositoryWrapper;
+          this.gsimWritePlatformService = gsimWritePlatformService;
+     }
+
+     /*
+      * Guaranteed to throw an exception no matter what the data integrity issue
+      * is.
+      */
+     private void handleDataIntegrityIssues(final JsonCommand command, final Throwable realCause, final Exception dve) {
+
+          final StringBuilder errorCodeBuilder = new StringBuilder("error.msg.")
+                    .append(SavingsApiConstants.SAVINGS_ACCOUNT_RESOURCE_NAME);
+
+          if (realCause.getMessage().contains("sa_account_no_UNIQUE")) {
+               final String accountNo = command.stringValueOfParameterNamed("accountNo");
+               errorCodeBuilder.append(".duplicate.accountNo");
+               throw new PlatformDataIntegrityException(errorCodeBuilder.toString(),
+                         "Savings account with accountNo " + accountNo + " already exists", "accountNo", accountNo);
+
+          } else if (realCause.getMessage().contains("sa_externalid_UNIQUE")) {
+
+               final String externalId = command.stringValueOfParameterNamed("externalId");
+               errorCodeBuilder.append(".duplicate.externalId");
+               throw new PlatformDataIntegrityException(errorCodeBuilder.toString(),
+                         "Savings account with externalId " + externalId + " already exists", "externalId", externalId);
+          }
+
+          errorCodeBuilder.append(".unknown.data.integrity.issue");
+          logger.error("{}", dve.getMessage(), dve);
+          throw new PlatformDataIntegrityException(errorCodeBuilder.toString(),
+                    "Unknown data integrity issue with savings account.");
+     }
+
+     @Transactional
+     @Override
+     public CommandProcessingResult submitGSIMApplication(final JsonCommand command) {
+
+          CommandProcessingResult result = null;
+
+          JsonArray gsimApplications = command.arrayOfParameterNamed("clientArray");
+
+          for (JsonElement gsimApplication : gsimApplications) {
```

Comment:
@rrpawar96  you may want to create an object and lock this method until it completes processing. If you are able to do it, you can remove applicationID param altogether.